cow_instrument/DetectorInfo.h

```diff
@@ -141,19 +156,60 @@ bool DetectorInfo<InstTree>::isMonitor(size_t detectorIndex) const {
   return m_isMonitor.const_ref()[detectorIndex];
 }
 
-template <typename InstTree> void DetectorInfo<InstTree>::initL2() {
+template <typename InstTree> void DetectorInfo<InstTree>::initL1() {
+
+  // Loop over all detector indexes. We will have a path for each.
   for (size_t detectorIndex = 0; detectorIndex < m_nDetectors;
        ++detectorIndex) {
-    const Detector &det = m_instrumentTree->getDetector(detectorIndex);
-    auto detPos = det.getPos();
 
-    /*
-     * Long-term is is not right. We cannot assume that the L2 path is simply
-     * sample
-     * to detector.
-     */
+    size_t i = 0;
+    const Path &path = (*m_l1Paths)[detectorIndex];
+    double l1 = 0;
+    // For each detector-l1-path calculate the total neutronic length
+    if (path.size() > 0) {
+      l1 += m_instrumentTree->getPathComponent(path[i]).length();
+      for (i = 1; i < path.size(); ++i) {
+        const PathComponent &current =
+            m_instrumentTree->getPathComponent(path[i]);
+        const PathComponent &previous =
+            m_instrumentTree->getPathComponent(path[i - 1]);
+        l1 += distance(current.entryPoint(), previous.exitPoint());
+        l1 += current.length();
+      }
+    }
+
+    (*m_l1)[detectorIndex] = l1;
+  }
+}
+
+template <typename InstTree> void DetectorInfo<InstTree>::initL2() {
+
+  // Loop over all detector indexes. We will have a path for each.
+  for (size_t detectorIndex = 0; detectorIndex < m_nDetectors;
+       ++detectorIndex) {
 
-    (*m_l2)[detectorIndex] = distanceToSample(detPos);
+    const Detector &det = m_instrumentTree->getDetector(detectorIndex);
+    auto detectorPos = det.getPos();
+    size_t i = 0;
+    const Path &path = (*m_l2Paths)[detectorIndex];
+    double l2 = 0;
+    if (path.size() > 0) {
+      l2 += m_instrumentTree->getPathComponent(path[i]).length();
+      // For each detector-l2-path calculate the total neutronic length
+      for (i = 1; i < path.size(); ++i) {
+        const PathComponent &current =
+            m_instrumentTree->getPathComponent(path[i]);
+        const PathComponent &previous =
+            m_instrumentTree->getPathComponent(path[i - 1]);
+        l2 += distance(current.entryPoint(), previous.exitPoint());
+        l2 += current.length();
+      }
+      l2 +=
+          distance(m_instrumentTree->getPathComponent(path[i - 1]).exitPoint(),
+                   detectorPos);
+    }
```

Comment:
I think this is more clear and I should try to resolve this. I don't know of any instrument that has a zero `L2` or detector pos = sample pos.